torch/csrc/jit/codegen/cuda/parallel_dimension_map.cpp

```diff
@@ -194,18 +196,50 @@ void ParallelDimensionMap::populateDimensionMapWithMultipleCASet(
   }
 }
 
-kir::Val* ParallelDimensionMap::get(ParallelType pt) const {
-  TORCH_INTERNAL_ASSERT(isParallelTypeThread(pt), "Invalid ParallelType: ", pt);
-  // Disable simplification of warp padded dimensions at
-  //  query time for now. Could extend this map to support
-  //  padded dimensions.
-  bool has_active_lower = GpuLower::current() != nullptr;
-  if (has_active_lower) {
-    auto& warp_info = GpuLower::current()->getWarpPaddedParallelInfo();
-    if (pt == ParallelType::TIDx && warp_info.is_tidx_padded) {
-      return kir::NamedScalar::getParallelDim(pt);
+void ParallelDimensionMap::adjustMappingsForWarpPadding() {
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  // If TIDx is padded to a multiple of the warp size, mark it as
+  // non-exact.
+
+  auto& warp_info = gpu_lower->getWarpPaddedParallelInfo();
+  if (!warp_info.is_tidx_padded) {
+    return;
+  }
+
+  const auto tidx_pt = ParallelType::TIDx;
+
+  // If the dimension of TIDx is actually a multple of the warp size
+  // before padding, it can be left as exact
+  if (isExact(tidx_pt)) {
```

Comment:
I guess one part I needed to confirm:
```
would it hurt predicate generation if `isExact` returns false but it is actually exact? 
```
So if we have a warp size of say, 33 and only one IterDomain parallelized on TIDx, then it'd still look exact to me but would be "exactly" 64 instead of 33.

I guess another subtle example is if we have two constants {61,62} and then it'd not be exact without padding but becomes exact after padding.

Reply:
I guess it'd be helpful to also confirm if I'm understanding `isExact` semantics correctly:

My understanding is `isExact` is true if either:
1. There is one compile-time constant, or multiple equal compile-time constants bound to this parallel type
or
2. There is one variable/runtime integer or one set of `multiple provably equal runtime integers and compile-time ones` bound to this parallel type.

And this check is compile time best effort, i.e. `isExact == true` means we can prove at compile time either of the above 2 is true and `isExact == false` means we did not prove it but not necessarily they're not true.

Reply:
```
My understanding is isExact is true if either:

There is one compile-time constant, or multiple equal compile-time constants bound to this parallel type
or
There is one variable/runtime integer or one set of multiple provably equal runtime integers and compile-time ones bound to this parallel type.
And this check is compile time best effort, i.e. isExact == true means we can prove at compile time either of the above 2 is true and isExact == false means we did not prove it but not necessarily they're not true.
```

That is correct.

The exactness in ParallelDimensionMap means that for all IterDomains that use the same ParallelType are proven to have the same extent as the dimension of the parallel type. If a ParallelType is not exact, some IterDomains have smaller extents than the threading dimension. So, in the latter case, we need to have a predicate like `threadIdx.x < some_iter_domain->extent()`. 

```
would it hurt predicate generation if `isExact` returns false but it is actually exact? 
```

Yes. It would end up with a predicate that's not actually necessary.

```
So if we have a warp size of say, 33 and only one IterDomain parallelized on TIDx, then it'd still look exact to me but would be "exactly" 64 instead of 33.
```

This is not considered exact as the parallel dimension is 64, while the IterDomain extent is 33, and I think that's how it should be defined. Warp padding does not change the size of tensors, but rather it just extends the number of threads, so 
predication should be necessary.

Hope this answers your questions.

Reply:
Oh I see. The point I was missing was that a parallel dimension is also not exact if it is to be set differently from the extents themselves, like in warp padding, even if the extents themselves are known. I think it all makes sense. Thanks for the clarifications 👍